File: scala/src/main/scala/Account.scala
package bank

import bank.domain._
import bank.printer._
import java.util.Date

class Account(val transactions: Transactions) {

  def deposit(amount: Double)(implicit clock: Clock): Unit = transactions.record(new Deposit(amount, clock.date))

  def withdraw(amount: Double)(implicit clock: Clock): Unit = transactions.record(new Withdraw(amount, clock.date))

  def printStatement(printer: Printer): String = printer.visit(this) + transactions.accept(printer)

}

File: java/src/main/java/bank/Account.java
package bank;

import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.StringJoiner;
import java.util.stream.Collectors;

public class Account {

	private Clock clock;
	private Transactions transactions = new InMemoryTransactions();
	
	public void deposit(float amount) {
		transactions.deposit(amount, clock.now());
	}

	public void withdraw(float amount) {
		transactions.withdraw(amount, clock.now());
	}

	public String printStatement() {
		List<String> rows = transactions.map(this::toRow);
		return rows.stream().collect(Collectors.joining("\n"));
	}
	
	public String toRow(Transaction transaction){
		StringJoiner joiner = new StringJoiner(" || ");
		return joiner.add(getDateString(transaction))
				.add(transaction.getCredit().map(String::valueOf).orElse(""))
				.add(transaction.getDebit().map(String::valueOf).orElse(""))
				.toString();
	}

	private String getDateString(Transaction transaction) {
		return transaction.getOperationDate().format(DateTimeFormatter.ofPattern("dd/MM/yyyy"));
	}
	
}
